### PR TITLE
Feat/73 maze chapter

### LINKED ISSUE #73: Feature: maze game chapter
- new api for maze-game-chapter and maze-game-chapter-session

### PR DESCRIPTION
**Pull request by quochieu586**
- add
1. Maze chapter, chapter session model
2. Maze chapter, chapter session controller
3. Maze chapter, chapter session service
- New api (view in google sheet)
- close #73

### CODE REVIEW

File: src/controllers/maze_chapter_session.controller.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                );
+
+            res.composer.success(result);
+        } catch (error) {
+            console.log(error);
+            res.composer.badRequest(error.message);
+        }
+    }
+
+    async startSession(req: Request, res: Response) {
```
Comment (tomgermanyyyy) [inline src/controllers/maze_chapter_session.controller.ts:47]:
```suggestion
    async createNewMazeSession(req: Request, res: Response) {

```

File: src/controllers/maze_chapter_session.controller.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            );
+
+            res.composer.success(result);
+        } catch (error) {
+            console.log(error);
+            res.composer.badRequest(error.message);
+        }
+    }
+
+    async getSession(req: Request, res: Response) {
```
Comment (tomgermanyyyy) [inline src/controllers/maze_chapter_session.controller.ts:67]:
```suggestion
    async getChapterSession(req: Request, res: Response) {

```

File: src/controllers/maze_session.controller.ts [hunk 3]
```diff
@@ -21,17 +21,19 @@ export class MazeSessionController extends Controller {
         this.router.post('/', this.createSession.bind(this));
         this.router.get('/current', this.getCurrentSession.bind(this));
         this.router.post('/:id/move', this.submitSingleMove.bind(this));
+        this.router.post('/:id/moves', this.submitMultipleMove.bind(this));
```
Comment (tomgermanyyyy) [inline src/controllers/maze_session.controller.ts:24]:
```suggestion
        this.router.post('/:id/moves', this.submitMultipleMoves.bind(this));

```

File: src/controllers/maze_session.controller.ts [hunk 4]
```diff
@@ -21,17 +21,19 @@ export class MazeSessionController extends Controller {
...
         this.router.post('/:id/end', this.endSession.bind(this));
+        this.router.get('/:id/score', this.getRoundScore.bind(this));
     }
 
     async createSession(req: Request, res: Response) {
         try {
             const userId = new Types.ObjectId(req.tokenMeta.userId);
 
-            const session = await this.mazeService.startSession(userId);
+            const session = await this.mazeService.startSession(userId, 1);
```
Comment (tomgermanyyyy) [inline src/controllers/maze_session.controller.ts:36]:
```suggestion
            const session = await this.mazeService.createSession(userId, 1);

```

File: src/controllers/maze_session.controller.ts [hunk 5]
```diff
@@ -137,4 +139,38 @@ export class MazeSessionController extends Controller {
             res.composer.badRequest(error.message);
         }
     }
+
+    async submitMultipleMove(req: Request, res: Response) {
+        try {
+            const sessionId = new mongoose.Types.ObjectId(req.params.id);
+            const userId = new Types.ObjectId(req.tokenMeta.userId);
+            const moves: string[] = req.body.moves;
+            const useHelp: boolean = req.body.isEnableAnimation;
```
Comment (tomgermanyyyy) [inline src/controllers/maze_session.controller.ts:148]:
```suggestion
            const canUseHelp: boolean = req.body.isEnableAnimation;

```

File: src/services/maze.service.ts [hunk 6]
```diff
@@ -169,31 +246,29 @@ export class MazeService {
...
-                break;
+        //     case 's':
+        //         moveDirection = Direction.Down;
+        //         break;
 
-            case 'a':
-                moveDirection = Direction.Left;
-                break;
+        //     case 'd':
+        //         moveDirection = Direction.Right;
```
Comment (tomgermanyyyy) [inline src/services/maze.service.ts:260]: xoa comment

File: src/services/maze_chapter_session.service.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+@injectable()
+export class MazeChapterSessionService {
+    constructor(@inject(ServiceType.Maze) private mazeService: MazeService) {}
+
+    async startChapterSession(
+        userId: Types.ObjectId,
+        chapterLevel: number,
+    ): Promise<MazeGameChapterSessionDocument> {
+        var isNewChapter = false;
```
Comment (tomgermanyyyy) [inline src/services/maze_chapter_session.service.ts:31]: no var

**Review by tomgermanyyyy — CHANGES_REQUESTED** [review]

File: src/controllers/maze_chapter_session.controller.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    constructor(
+        @inject(ServiceType.Auth) private authService: AuthService,
+        @inject(ServiceType.MazeChapterSession)
+        private mazeChapterSessionService: MazeChapterSessionService,
+    ) {
+        super();
+
+        this.router.all('*', this.authService.authenticate());
+        this.router.post('/', this.createNewChapterSession.bind(this));
+        this.router.post('/:id/start', this.startSession.bind(this));
```
[the remark this answers is not in the document]
Comment (quochieu586) [reply src/controllers/maze_chapter_session.controller.ts:23]: Dạ ròi

File: src/controllers/maze_chapter_session.controller.ts (hunk 1 above)
Comment (quochieu586) [reply src/controllers/maze_chapter_session.controller.ts:47]: Dạ ròi :3

File: src/controllers/maze_chapter_session.controller.ts (hunk 2 above)
Comment (quochieu586) [reply src/controllers/maze_chapter_session.controller.ts:67]: Dạ ròi

File: src/controllers/maze_session.controller.ts (hunk 4 above)
Comment (quochieu586) [reply src/controllers/maze_session.controller.ts:36]: Dạ ròi

File: src/controllers/maze_session.controller.ts (hunk 5 above)
Comment (quochieu586) [reply src/controllers/maze_session.controller.ts:148]: Dạ ròi

File: src/services/maze.service.ts (hunk 6 above)
Comment (quochieu586) [reply src/services/maze.service.ts:260]: Dạ ròi :3

File: src/services/maze_chapter_session.service.ts (hunk 7 above)
Comment (quochieu586) [reply src/services/maze_chapter_session.service.ts:31]: Dạ ròi

File: src/controllers/maze_session.controller.ts (hunk 3 above)
Comment (quochieu586) [reply src/controllers/maze_session.controller.ts]: Dạ ròi

**Review by adminfessior — DISMISSED** [review]

**Review by tomgermanyyyy — DISMISSED** [review]

**Review by tomgermanyyyy — APPROVED** [review]

**Review by tomgermanyyyy — APPROVED** [review]

**Pull request merged**